src/container/srv_container.c

```diff
@@ -369,17 +369,99 @@ enum nhandles_op {
 	NHANDLES_PRE_DECREMENT
 };
 
+/** Number of handles hash table record - cache rdb num_handles within one tx closing a batch. */
+struct nhandles_ht_rec {
+	uuid_t		nhr_cuuid;
+	uint32_t	nhr_nhandles;
+	uint32_t	nhr_ref;
+	d_list_t	nhr_hlink;
+};
+
+static inline struct nhandles_ht_rec *
+nhandles_rec_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct nhandles_ht_rec, nhr_hlink);
+}
+
+static void
+nhandles_ht_rec_addref(struct d_hash_table *ht, d_list_t *link)
+{
+	struct nhandles_ht_rec *rec = nhandles_rec_obj(link);
+
+	rec->nhr_ref++;
+	D_DEBUG(DB_TRACE, "rec=%p, incremented nhr_ref to %u\n", rec, rec->nhr_ref);
+}
+
+static bool
+nhandles_ht_rec_decref(struct d_hash_table *ht, d_list_t *link)
+{
+	struct nhandles_ht_rec *rec = nhandles_rec_obj(link);
+
+	rec->nhr_ref--;
+	D_DEBUG(DB_TRACE, "rec=%p, decremented nhr_ref to %u\n", rec, rec->nhr_ref);
+	return (rec->nhr_ref == 0);
+}
+
+static bool
+nhandles_ht_cmp_keys(struct d_hash_table *ht, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct nhandles_ht_rec *rec = nhandles_rec_obj(rlink);
+
+	return uuid_compare(key, rec->nhr_cuuid) == 0;
+}
+
+static void
+nhandles_ht_rec_free(struct d_hash_table *htable, d_list_t *link)
+{
+	struct nhandles_ht_rec *rec = nhandles_rec_obj(link);
+
+	D_ASSERT(d_hash_rec_unlinked(&rec->nhr_hlink));
+	D_DEBUG(DB_MD, "Free rec=%p\n", rec);
+	D_FREE(rec);
+}
+
+static d_hash_table_ops_t nhandles_hops = {
+	.hop_key_cmp = nhandles_ht_cmp_keys,
+	.hop_rec_addref = nhandles_ht_rec_addref,
+	.hop_rec_decref = nhandles_ht_rec_decref,
+	.hop_rec_free = nhandles_ht_rec_free,
+};
 
-/* Get container number of open handles, if the key exists in rdb ; optionally update based on op */
 static int
-get_nhandles(struct rdb_tx *tx, struct cont *cont, enum nhandles_op op, uint32_t *nhandles)
+nhandles_ht_create(struct d_hash_table *nht)
 {
-	uint32_t	result = 0;
-	d_iov_t		value;
-	bool		do_update = true;
-	int		rc;
+	return d_hash_table_create_inplace(D_HASH_FT_NOLOCK, 6 /* bits */, NULL /* priv */,
+					   &nhandles_hops, nht);
+}
 
-	d_iov_set(&value, &result, sizeof(result));
+static int
+nhandles_ht_destroy(struct d_hash_table *nht)
+{
+	int rc;
+
+	rc = d_hash_table_destroy_inplace(nht, true);
+	if (rc)
+		D_ERROR("d_hash_table_destroy_inplace() failed, "DF_RC"\n", DP_RC(rc));
+	return rc;
+}
+
+/* Get, optionally update container number of handles. Use "nhandles cache" (nhc) if provided. */
+static int
+get_nhandles(struct rdb_tx *tx, struct d_hash_table *nhc, struct cont *cont, enum nhandles_op op,
+	     uint32_t *nhandles)
+{
+	uint32_t		result = 0;
+	d_iov_t			value;
+	uint32_t		lookup_val;		/* value from DRAM cache or rdb */
+	struct nhandles_ht_rec *rec = NULL;
+	struct nhandles_ht_rec *rec_inserted = NULL;
+	bool			do_update = true;
+	int			rc;
+
+	/* Fetch nhandles value from rdb first. Might be inaccurate if the tx has updated already.
+	 * However, this is necessary e.g., just for the interop check.
+	 */
```

Comment:
[Question] Would it be possible to perform the interop check using the pool global version? If yes, the algorithm could have a more intuitive order between the cache lookup and the DB lookup:

    if layout without nhandles
        goto out

    if with cache
        look up cache // only involves memory accesses
    if without cache or not found
        look up DB    // may involve SSD reads

    modify value

    if with cache
        update cache
    update DB

And with "MD on SSD phase 2," the `rdb_tx_lookup` call may involve SSD reads.

Reply:
done in commit [2760feb](https://github.com/daos-stack/daos/pull/11736/commits/2760feb7f9bdf8721c07373e6136de31119ee7ca)